Adds remirror functionality

MCterra10 opened the pull request:
I literally learnt go just to fix this shit.

MCterra10 committed:
adds remirror functionality

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
fixes syntax error
there's probably more but that's one

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
code restructure

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
fixes more errors

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
fixes moar bugs

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
more bug fixes again

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
bruh moment

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
Fix bug

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
fix it or else

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
help I'm stupid

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
testing something

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
facepalm

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
Adds golint comments, adds remirror method, shifts stuff around a bit

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
fix logical error and other issue

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 commented:
worth noting that I moved labeling the PR to a webhook and added a force push to the script to remirror properly

MCterra10 committed:
should probably keep this for regular mirrors

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 commented:
hurry up you dumbass :b:ippie maintainers

j-mie commented:
I don't do anything for Hippie anymore or play SS13 but I've asked them to add me back to the repo so I can review and merge this

j-mie commented on prmirror.go:
```diff
@@ -43,20 +43,32 @@ func (p PRMirror) HandleEvent(event *github.Event) {
-	p.HandlePREvent(&prEvent)
-	p.Database.AddEvent(event.GetID())
+		p.HandlePRComment(&prComment)
+		p.Database.AddEvent(event.GetID())
```
I'm not sure if these IDs are the same kind of ID for both PullRequestEvents and CommentEvents are they? If they're the same ID then this is fine but if not you're going to need to prepend `pr` and `comment` to the IDs to stop collisions between IDs and PRs

MCterra10 replied:
hmm, I can do that, shouldn't be too hard to alter the method

MCterra10 replied:
done

j-mie commented on prmirror.go:
```diff
@@ -78,12 +90,44 @@ func (p PRMirror) HandlePREvent(prEvent *github.PullRequestEvent) {
 		if err != nil {
 			log.Errorf("Error while creating a new PR: %s\n", err.Error())
 		} else {
-			p.AddLabels(prID, []string{"Upstream PR Merged"})
```
How come you're removing this?

MCterra10 replied:
That's my bad, I removed it because it's handled by our webhook system. If it's an issue it can be added back.

MCterra10 replied:
made it into a config option

j-mie commented on prmirror.go:
```diff
@@ -78,12 +90,44 @@ func (p PRMirror) HandlePREvent(prEvent *github.PullRequestEvent) {
+			return
+		}
+		body := pr.GetBody()
+		temp := strings.Split(body, "/")
```
what is this even doing. It looks like you're getting the upstream PullRequest ID from the downstream PR maybe? If so there's definitely a better way of doing that

j-mie replied:
Looks like we store a map of Upstream/Downstream using `StoreMirror` - you can probably implement a way to get the Upstream PR ID from the Downstream PR ID using that

MCterra10 replied:
literally no experience in go other than what I found online in brief tutorials, so that's really beyond my scope I'm afraid

MCterra10 replied:
I added some more error handling for this, other than that I don't really know how to improve it further, given my experience with go.

MCterra10 commented:
oh wow someone answered, okay I'll address the reviews when I get some time

MCterra10 committed:
unique IDs, adds config for labels

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
we now strip github autolinks
this is to ensure the PRs don't accidentally close
our own issues when merged.

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
Add some more files to .gitignore

Signed-off-by: Terra <terra@mcterra.id.au>

MCterra10 committed:
Update README.md

Neotwr commented:
@MCterra10 why you close

MCterra10 commented:
@Neo1w maintainer wasn't responding

MCterra10 commented:
If you want the code changes, just fork from austation's PR Mirror repo.

Neotwr commented:
fair
